var/spack/repos/builtin/packages/apcomp/package.py

```diff
@@ -0,0 +1,185 @@
+# Copyright 2013-2020 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+
+from spack import *
+import os
+import socket
+import llnl.util.tty as tty
+
+
+def cmake_cache_entry(name, value, vtype=None):
+    """
+    Helper that creates CMake cache entry strings used in
+    'host-config' files.
+    """
+    if vtype is None:
+        if value == "ON" or value == "OFF":
+            vtype = "BOOL"
+        else:
+            vtype = "PATH"
+    return 'set({0} "{1}" CACHE {2} "")\n\n'.format(name, value, vtype)
+
+
+class Apcomp(Package):
+    """A multi use-case image compositor"""
+
+    homepage = 'https://github.com/Alpine-DAV/ap_compositor'
+    git      = 'https://github.com/Alpine-DAV/ap_compositor.git'
+    url      = "https://github.com/Alpine-DAV/ap_compositor/releases/download/v0.0.1/apcomp-v0.0.1.tar.gz"
+
+    maintainers = ['mclarsen', 'cyrush']
+
+    version('master', branch='master', submodules='True')
+    version('0.0.1', sha256="cbf85fe58d5d5bc2f468d081386cc8b79861046b3bb7e966edfa3f8e95b998b2", preferred=True)
+
+    variant('openmp', default=True, description='Build with openmp support')
+    variant('mpi', default=True, description='Build with openmp support')
+    variant('shared', default=True, description='Build Shared Library')
+
+    depends_on('cmake@3.9:', type='build')
+    depends_on("mpi", when="+mpi")
+
+    root_cmakelists_dir = 'src'
+
+    def install(self, spec, prefix):
+        """
+        Build and install APComp
+        """
+        with working_dir('spack-build', create=True):
+            py_site_pkgs_dir = None
+            host_cfg_fname = self.create_host_config(spec,
+                                                     prefix,
+                                                     py_site_pkgs_dir)
+            cmake_args = []
+            # if we have a static build, we need to avoid any of
+            # spack's default cmake settings related to rpaths
+            # (see: https://github.com/LLNL/spack/issues/2658)
+            if "+shared" in spec:
+                cmake_args.extend(std_cmake_args)
+            else:
+                for arg in std_cmake_args:
+                    if arg.count("RPATH") == 0:
+                        cmake_args.append(arg)
+            cmake_args.extend(["-C", host_cfg_fname, "../src"])
+            print("Configuring APComp...")
+            cmake(*cmake_args)
+            print("Building APComp...")
+            make()
+            print("Installing APComp...")
+            make("install")
+            # install copy of host config for provenance
+            install(host_cfg_fname, prefix)
+
+    def create_host_config(self, spec, prefix, py_site_pkgs_dir=None):
```

Comment:
Is there any reason we can't extend `CMakePackage` instead of adding all this custom logic?

Reply:
we need to capture the relevant cmake config values for provenance, debugging, and to allow folks to build when developing against an existing set of spack-built tpls.

Another perspective: The magic folded into the CMakePackage via inheritance shifts over time -- it's hard to keep up with choices that being made as things evolve and are convolved together without extensive vetting.